Site Editor Tracking - Template Part creation / replacement flows.

Addison-Stavlo opened the pull request:
#### Changes proposed in this Pull Request

Adds tracking for various template part flows:

* Creating a new template part via the placeholder.  `wpcom_block_editor_create_template_part` with properties for `variation_slug` and `content` if applicable.  If the user has selected a pattern in the creation flow, `content` will correspond to that patterns markup.  Otherwise, the `content` property will be absent as the user selected to create a blank template part.
* Choosing an existing template part via the placeholder.  `wpcom_block_editor_template_part_choose_existing` with properties of `variation_slug` and `template_part_id` (such as "pub/tt1-blocks//header") of the selected template part.
* Replacing a template part via the block inserter.  `wpcom_block_editor_template_part_replace` will have 2 properties corresponding to the replaced item (`replaced_template_part_id` and `replaced_variation_slug`) and 2 properties corresponding to the newly selected item (`template_part_id` and `variation_slug`).

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Follow steps to enable tracking debugging at PCYsg-nrf-p2
* Open the site editor.
* Add a template part block ( "Header" or "Footer")
* Using the placeholder, select "New header" (or footer).  
* Select "choose" from the pattern selector tool.  Verify the `wpcom_block_editor_create_template_part` event fires with the expected properties as noted above.
* Repeat the above but instead of selecting to "choose" a pattern, select "Start blank".  Verify the above event fires and there is no `content` property.
* Add another template part block.  This time, instead of selecting "New template part", select "choose existing" and select an existing template part from the dropdown.  Verify that `wpcom_block_editor_template_part_choose_existing` event fires with the expected properties listed above.
* Next, select an existing and instantiated template part in the editor (this could be one you created in a previous step).  With the parent template part block selected, select "replace" in the block toolbar:
![Screen Shot 2021-06-28 at 2 12 08 PM](https://user-images.githubusercontent.com/28742426/123690161-0f2c9600-d822-11eb-9faf-0eec09cc021b.png)
* Select a template part to replace with (note, another template part of this area/variation must exist).  Verify the `wpcom_block_editor_template_part_replace` event fires with the expected properties `replaced_template_part_id`, `replaced_variation_slug`, `template_part_id`, and `variation_slug`.

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/issues/53410

Addison-Stavlo committed:
track creation

Addison-Stavlo committed:
track replace and choose existing

Addison-Stavlo commented on apps/wpcom-block-editor/src/wpcom/features/tracking.js:
```diff
@@ -391,6 +391,21 @@ const trackListViewToggle = ( isOpen ) => {
+		if ( document.querySelector( '.edit-site-template-part-converter__modal' ) ) {
+			tracksRecordEvent( 'wpcom_block_editor_convert_to_template_part', {
+				variation_slug: record.area !== 'uncategorized' ? record.area : undefined,
+				content: record.content,
```
Note for the convert part this `content` should correspond to a flattened list of blocks as laid out in https://github.com/Automattic/wp-calypso/pull/53592.  However, for the `*_create_template_part` event below using record.content makes sense as this will correspond directly with a block pattern's markup, is finite, and is not a users customized and written content.

david-szabo97 commented:
Works well so far! Looking forward to the tests 😄

Addison-Stavlo committed:
handle conflicts

Addison-Stavlo committed:
why wont text locator work?

Addison-Stavlo commented on test/e2e/specs/specs-wpcom/wp-calypso-gutenberg-site-editor-tracking-spec.js:
```diff
@@ -634,6 +634,33 @@ describe( `[${ host }] Calypso Gutenberg Site Editor Tracking: (${ screenSize })
+			it( 'Tracks "wpcom_block_editor_create_template_part', async function () {
+				const editor = await SiteEditorComponent.Expect( this.driver );
+				await editor.addBlock( 'Header', 'template-part\\/header', 'Block: Template Part' );
+				const createNewHeaderLocator = driverHelper.createTextLocator(
+					By.css( '.wp-block-template-part:last-of-type .components-placeholder button' ),
+					'New header'
+				);
+				await driverHelper.clickWhenClickable( this.driver, createNewHeaderLocator );
```
Im not sure why this isnt working.  I can verify the css selector targets the buttons in the placeholder.  And I can verify the button's text is 'New header' ....  but nope.

Addison-Stavlo committed:
fix create test

Addison-Stavlo committed:
tests for choose and replace

Addison-Stavlo committed:
rebase trunk

Addison-Stavlo committed:
need more trunk

Addison-Stavlo committed:
even more trunk

Addison-Stavlo committed:
add more checks, tests are working

david-szabo97 commented:
E2E fails on mobile because the global style remains open after the previous tests.

Rebase after https://github.com/Automattic/wp-calypso/pull/54195 and see if it works

WunderBart commented on test/e2e/specs/specs-wpcom/wp-calypso-gutenberg-site-editor-tracking-spec.js:
```diff
@@ -737,6 +737,132 @@ describe( `[${ host }] Calypso Gutenberg Site Editor Tracking: (${ screenSize })
+
+				const { variation_slug, template_part_id } = chooseEvents[ 0 ][ 1 ];
+				// Check the event props, assert id.length > 2 since the format is `{theme}//{slug}`.
+				assert(
+					variation_slug === 'header' &&
+						typeof template_part_id === 'string' &&
+						template_part_id.length > 2
+				);
```
nit: Might be easier to follow debugging-wise when checked via separate assertions:
```suggestion
				assert( variation_slug === 'header' );
				assert( typeof template_part_id === 'string' );
				assett( template_part_id.length > 2 );
```

WunderBart reviewed (approved):
Looks good E2E-wise! Thanks! 🚀

Addison-Stavlo committed:
Update test/e2e/specs/specs-wpcom/wp-calypso-gutenberg-site-editor-tracking-spec.js

Co-authored-by: Bart Kalisz <bartlomiej.kalisz@gmail.com>

Addison-Stavlo committed:
Merge branch 'trunk' into add/create-template-part-tracking

Addison-Stavlo committed:
fix undo to setup test on mobile

Addison-Stavlo committed:
clear block selection to start template part tests

Addison-Stavlo commented:
Rebased and fixed the tests on mobile.  However, now when I run desktop it looks like `runInCanvas` errors because the dotcom bug report button gets in the way:

![Screen Shot 2021-07-06 at 4 30 39 PM](https://user-images.githubusercontent.com/28742426/124663109-a96e8880-de77-11eb-8f5b-38ed66fefd32.png)

🤦‍♀️

david-szabo97 commented:
> Rebased and fixed the tests on mobile. However, now when I run desktop it looks like `runInCanvas` errors because the dotcom bug report button gets in the way:
> 
> ![Screen Shot 2021-07-06 at 4 30 39 PM](https://user-images.githubusercontent.com/28742426/124663109-a96e8880-de77-11eb-8f5b-38ed66fefd32.png)
> 
> 🤦‍♀️

Welcome to the club 😄  I had to rewrite a whole test because of this 😬

WunderBart commented:
> > Rebased and fixed the tests on mobile. However, now when I run desktop it looks like `runInCanvas` errors because the dotcom bug report button gets in the way:
> > ![Screen Shot 2021-07-06 at 4 30 39 PM](https://user-images.githubusercontent.com/28742426/124663109-a96e8880-de77-11eb-8f5b-38ed66fefd32.png)
> > 🤦‍♀️
> 
> Welcome to the club 😄 I had to rewrite a whole test because of this 😬

Oh hai, I've even filed an issue for this 😄 https://github.com/Automattic/wp-calypso/issues/50840

Addison-Stavlo commented:
LOL!  Well at least we are all in this club together. 👯  😆

Addison-Stavlo committed:
reload before creation/replacement tests

Addison-Stavlo commented:
The e2e tests seem to be working on both desktop and mobile now.  I added a step to reload the editor for this `describe` block, so it gets rid of all the garbage from inserting blocks/patterns and other goo from previous tests.  That seems to allow us to avoid conflicting with the bug report button here.

david-szabo97 reviewed (approved):
Works as expected!
